src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -2163,7 +2193,8 @@ private unsafe void DispatchToAllListeners(int eventId, EventWrittenEventArgs ev
             for (EventDispatcher? dispatcher = m_Dispatchers; dispatcher != null; dispatcher = dispatcher.m_Next)
             {
                 Debug.Assert(dispatcher.m_EventEnabled != null);
-                if (eventId == -1 || dispatcher.m_EventEnabled[eventId])
```

Comment:
This dispatcher.m_EventEnabled[eventId] appears like the intended pre-existing mechanism to do per-EventListener filtering. Before we introduce a new mechanism I'd want to understand:
a) why doesn't this one work?
b) are we better off introducing a new state tracking mechanism instead of fixing this one? (my hunch is 'no' but I don't understand the issue yet)

Reply:
Thanks for the feedback Noah. I did look into `dispatcher.m_EventEnabled[eventId]` and the reason it doesn't work is because of EventSources that are self-describing. Events from self-describing EventSources have EventIds of -1, and these aren't updated as a result. So we end up just passing through everything via the `eventID == -1` check.

Reply:
Ah, that makes much more sense, thanks! A couple thoughts:

1. Please add to the comments why SelfDescribing events need the alternate filtering mechanism : )
2. We already have the EventDispatcher object inside the EventSource m_Dispatchers list that corresponds to a particular (EventSource, EventListener) tuple. Rather than create a new Dictionary of EventListenerEventState objects I suspect we could re-use the pre-existing m_Dispatchers data structure? You could add a level and keywords field to the EventDispatcher. Aside from reducing the number of types and data structures this also appears to better align performance characteristics. WriteToAllListeners is part of the event writing hot path we want to optimize. Anything we can precompute in EnableEvents to create data structures that are cheaper to access on event writes, the better. Storing the data in Dispatcher eliminates the needing to do a Dictionary lookup to get the EventListenerEventState from the listener, probably saving ~20ns per-listener per-write.
3. Rather than doing a separate UpdateKeywordsAndLevels(), can we reuse the code paths that already occur inside EnableEvents()? It appears that EnableEvents() calls to SendCommand which ultimately goes to DoCommand(). In the body of that function is already logic to find the right dispatcher and update the event flags (via calls to EnableEventForDispatcher). It would fit pretty easily to update the keyword and level on the dispatcher at that point so that we are doing the self-describing and manifest based filter state caching side by side.

Reply:
Thanks for the suggestions :-) 

> Please add to the comments why SelfDescribing events need the alternate filtering mechanism : )

Done!

> Rather than create a new Dictionary of EventListenerEventState objects I suspect we could re-use the pre-existing m_Dispatchers data structure? 

I refactored the change to save the level/keyword combination in the EventDispatcher class instead, leaving EventListener as-is.  

> Rather than doing a separate UpdateKeywordsAndLevels(), can we reuse the code paths that already occur inside EnableEvents()? 

Removed these and moved the bookkeeping logic to DoCommand().